File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        start, curr, index = head, head, 1
        while index < left:
            start = curr
            curr = curr.next
            index += 1
        
        reverse, tail = None, curr
        
        while index >= left and index <= right:
            next = curr.next
            [curr.next, reverse] = [reverse, curr]
            curr = next
            index += 1
        
        start.next = reverse
        tail.next = curr
        
        return head if left > 1 else reverse
        

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseBetween(head: ListNode | null, left: number, right: number): ListNode | null {
    let curr = head, start = head, index = 1;
    while(index < left){
        start = curr;
        curr = curr.next;
        index++;
    }
    
    let reverse = null, tail = curr;
    
    while(index >= left && index <= right){
        let next = curr.next;
        [curr.next, reverse] = [reverse, curr];
        curr = next;
        index++;
    }
    
    start.next = reverse;
    tail.next = curr;
    
    return left > 1 ? head : reverse
}
